### ISSUE TITLE
[crash] double free detected in tcache 2

### ISSUE THREAD

**Issue by antoinemadec** [neovim/neovim#16287]
### Neovim version (nvim -v)

NVIM v0.6.0-dev+561-g03b805aee

### Vim (not Nvim) behaves the same?

Cannot tell as the plugin I am using does not support vim

### Operating system/version

Manjaro Linux

### Terminal name/version

alacritty 0.9.0 (fed349aa)

### $TERM environment variable

xterm-256color

### Installation

build from repo

### How to reproduce the issue

1. nvim -u vimrc foo.c
2. :CocFzfList symbols
3. press escape
4. :CocFzfList symbols
5. press escape
6. wait for a couple of seconds

vimrc:
```vim
call plug#begin('./plugins')
Plug 'neoclide/coc.nvim', {'branch': 'release'}
Plug 'junegunn/fzf', {'dir': '~/.fzf','do': './install --all'}
Plug 'junegunn/fzf.vim'
Plug 'antoinemadec/coc-fzf', {'commit': '2d9deb2a3b05a0c09f01edda0a4abd542cb3eebd'}
call plug#end()
```

foo.c:
```c
int main(int argc, char *argv[])
{
  return 0;
}
```

### Expected behavior

No crash.

### Actual behavior

Crash with message: 
```
free(): double free detected in tcache 2
{                                       Aborted (core dumped)
```

Coredump:
```
> coredumpctl dump 19062
           PID: 19062 (nvim)
           UID: 1000 (antoine)
           GID: 1000 (antoine)
        Signal: 6 (ABRT)
     Timestamp: Thu 2021-11-11 17:47:30 CET (1min 5s ago)
  Command Line: nvim -u vimrc foo.c
    Executable: /usr/local/bin/nvim
 Control Group: /user.slice/user-1000.slice/session-2.scope
          Unit: session-2.scope
         Slice: user-1000.slice
       Session: 2
     Owner UID: 1000 (antoine)
       Boot ID: 07ec2911b24b4666834e355378a3ff3c
    Machine ID: 2dcc56548c094a528229dd447dbc4b2a
      Hostname: antoine-inspiron137368
       Storage: /var/lib/systemd/coredump/core.nvim.1000.07ec2911b24b4666834e355378a3ff3c.19062.1636649250000000.zst (present)
     Disk Size: 1.0M
       Message: Process 19062 (nvim) of user 1000 dumped core.

                Found module linux-vdso.so.1 with build-id: 48639d17f4288910298f29eaecfe352a95f6c659
                Found module libnss_files.so.2 with build-id: 1a36dfc01d3a1010b2ee79766a24a8090a3266d5
                Found module libresolv.so.2 with build-id: c915c72668282861a813f7ea3c0780f37b681dc0
                Found module libkeyutils.so.1 with build-id: ac405ddd17be10ce538da3211415ee50c8f8df79
                Found module libkrb5support.so.0 with build-id: adf65240a4d2aba772d7a0772b4d015469934113
                Found module libcom_err.so.2 with build-id: eb61ef71c8b97846db759fb89a115405cff6dd30
                Found module libk5crypto.so.3 with build-id: eb8220b8f36675aac769450be4cb6bb7f97ec38a
                Found module libkrb5.so.3 with build-id: 72d26767c5cb1097db75a5f5bff88860233c902b
                Found module libgssapi_krb5.so.2 with build-id: e6e098ad51ce7bdd3dbe902d7b0f69a90f8a9e08
                Found module libtirpc.so.3 with build-id: 5bef2adfdee3df283f593b3e2d37b6dac405256a
                Found module ld-linux-x86-64.so.2 with build-id: 040cc3dd10461562f177df39e3be2f3704258c3c
                Found module libc.so.6 with build-id: 4b406737057708c0e4c642345a703c47a61c73dc
                Found module libgcc_s.so.1 with build-id: 7f8508bb914546ada778809b64b99d234337d835
                Found module libutil.so.1 with build-id: af6d8200064ee6e6a01252884d166b44f0a6121b
                Found module libm.so.6 with build-id: 2b8fd1f869ecab4e0b55e92f2f151897f6818acf
                Found module libdl.so.2 with build-id: 5abc547e7b0949f89f3c0e21ab0c8331a7440a8a
                Found module libnsl.so.3 with build-id: 059e45eb1f084e0bb3143f0c4c364a6ade7d50b1
                Found module libpthread.so.0 with build-id: 07c8f95b4f3251d08550217ad8a1f31066229996
                Found module librt.so.1 with build-id: 75484da2d6f1515189eefa076e0a40328834cd16
                Found module nvim with build-id: a7af43909d7d48517d02813abce3cfa9d17b93f7
                Stack trace of thread 19062:
                #0  0x00007f0b3e239d22 raise (libc.so.6 + 0x3cd22)
                #1  0x00007f0b3e223862 abort (libc.so.6 + 0x26862)
                #2  0x00007f0b3e27bd28 __libc_message (libc.so.6 + 0x7ed28)
                #3  0x00007f0b3e28392a malloc_printerr (libc.so.6 + 0x8692a)
                #4  0x00007f0b3e285408 _int_free (libc.so.6 + 0x88408)
                #5  0x00007f0b3e2889e8 __libc_free (libc.so.6 + 0x8b9e8)
                #6  0x0000563d76dd2e1b vars_clear_ext (nvim + 0x123e1b)
                #7  0x0000563d76e0ddf8 free_unref_funccal (nvim + 0x15edf8)
                #8  0x0000563d76dc941e garbage_collect (nvim + 0x11a41e)
                #9  0x0000563d76f07b27 os_inchar (nvim + 0x258b27)
                #10 0x0000563d76f94f0d state_enter (nvim + 0x2e5f0d)
                #11 0x0000563d76ecc56b normal_enter (nvim + 0x21d56b)
                #12 0x0000563d76d445ad main (nvim + 0x955ad)
                #13 0x00007f0b3e224b25 __libc_start_main (libc.so.6 + 0x27b25)
                #14 0x0000563d76d4582e _start (nvim + 0x9682e)

                Stack trace of thread 19063:
                #0  0x00007f0b3e2fb92e epoll_wait (libc.so.6 + 0xfe92e)
                #1  0x0000563d77043457 uv__io_poll (nvim + 0x394457)
                #2  0x0000563d7702fc34 uv_run (nvim + 0x380c34)
                #3  0x0000563d76fbe73e tui_main (nvim + 0x30f73e)
                #4  0x00007f0b3e548259 start_thread (libpthread.so.0 + 0x9259)
                #5  0x00007f0b3e2fb5e3 __clone (libc.so.6 + 0xfe5e3)
Refusing to dump core to tty (use shell redirection or specify --output).
```

**Comment by gpanders** [comment]
Thanks for the report and for including the backtrace, but we are going to need a better repro case. Installing four plugins doesn't satisfy the definition of "minimal".

**Comment by antoinemadec** [comment]
@gpanders I know, and I tried, but this is not easy.
The crash goes away if I reduce the number of local variables of my function ```coc_fzf#symbols#fzf_run```

I was able to run valgrind, which makes this crash go away:
```
==20701== Memcheck, a memory error detector
==20701== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==20701== Using Valgrind-3.17.0 and LibVEX; rerun with -h for copyright info
==20701== Command: nvim -u vimrc foo.c
==20701== Parent PID: 1210
==20701== 
==20701== Warning: noted but unhandled ioctl 0x5441 with no size/direction hints.
==20701==    This could cause spurious value errors to appear.
==20701==    See README_MISSING_SYSCALL_OR_IOCTL for guidance on writing a proper wrapper.
==20701== Invalid free() / delete / delete[] / realloc()
==20701==    at 0x484118B: free (vg_replace_malloc.c:755)
==20701==    by 0x22BE1A: vars_clear_ext (in /usr/local/bin/nvim)
==20701==    by 0x266DF7: free_unref_funccal (in /usr/local/bin/nvim)
==20701==    by 0x22241D: garbage_collect (in /usr/local/bin/nvim)
==20701==    by 0x360B26: os_inchar (in /usr/local/bin/nvim)
==20701==    by 0x3EDF0C: state_enter (in /usr/local/bin/nvim)
==20701==    by 0x32556A: normal_enter (in /usr/local/bin/nvim)
==20701==    by 0x19D5AC: main (in /usr/local/bin/nvim)
==20701==  Address 0x5140660 is 0 bytes inside a block of size 256 free'd
==20701==    at 0x484118B: free (vg_replace_malloc.c:755)
==20701==    by 0x1C1CF0: api_free_object (in /usr/local/bin/nvim)
==20701==    by 0x1C1C9D: api_free_object (in /usr/local/bin/nvim)
==20701==    by 0x1C1CF0: api_free_object (in /usr/local/bin/nvim)
==20701==    by 0x1C1C9D: api_free_object (in /usr/local/bin/nvim)
==20701==    by 0x1C1CF0: api_free_object (in /usr/local/bin/nvim)
==20701==    by 0x45B034: serialize_response.constprop.0 (in /usr/local/bin/nvim)
==20701==    by 0x32C789: request_event.lto_priv.0 (in /usr/local/bin/nvim)
==20701==    by 0x321575: parse_msgpack.lto_priv.0 (in /usr/local/bin/nvim)
==20701==    by 0x3219DC: receive_msgpack (in /usr/local/bin/nvim)
==20701==    by 0x267CA7: read_event (in /usr/local/bin/nvim)
==20701==    by 0x26BA15: invoke_read_cb (in /usr/local/bin/nvim)
==20701==  Block was alloc'd at
==20701==    at 0x483E6D5: malloc (vg_replace_malloc.c:379)
==20701==    by 0x30D3CF: xrealloc (in /usr/local/bin/nvim)
==20701==    by 0x1C8A50: copy_array (in /usr/local/bin/nvim)
==20701==    by 0x1C8910: copy_object (in /usr/local/bin/nvim)
==20701==    by 0x1C8A0B: copy_array (in /usr/local/bin/nvim)
==20701==    by 0x1C8910: copy_object (in /usr/local/bin/nvim)
==20701==    by 0x1C8716: copy_dictionary (in /usr/local/bin/nvim)
==20701==    by 0x1C88CA: copy_object (in /usr/local/bin/nvim)
==20701==    by 0x1C8A0B: copy_array (in /usr/local/bin/nvim)
==20701==    by 0x1C8910: copy_object (in /usr/local/bin/nvim)
==20701==    by 0x1C8716: copy_dictionary (in /usr/local/bin/nvim)
==20701==    by 0x1CABBF: api_metadata (in /usr/local/bin/nvim)
==20701== 
==20701== 
==20701== HEAP SUMMARY:
==20701==     in use at exit: 7,376,715 bytes in 39,995 blocks
==20701==   total heap usage: 212,152 allocs, 172,158 frees, 342,182,611 bytes allocated
==20701== 
==20701== LEAK SUMMARY:
==20701==    definitely lost: 0 bytes in 0 blocks
==20701==    indirectly lost: 0 bytes in 0 blocks
==20701==      possibly lost: 610,625 bytes in 12,083 blocks
==20701==    still reachable: 6,766,090 bytes in 27,912 blocks
==20701==         suppressed: 0 bytes in 0 blocks
==20701== Rerun with --leak-check=full to see details of leaked memory
==20701== 
==20701== For lists of detected and suppressed errors, rerun with: -s
==20701== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

Would valgrind be enough or do we still need to spend more time coming up with a minimal vimrc?

**Comment by antoinemadec** [comment]
@gpanders I managed to come up with a minimal repro:

1. nvim --clean
2. source test.vim

test.vim:
```vim
function! Foo() abort
  " no bug without s:
  " no bug without a:000
  let s:args = [a:000]

  " no bug if we set less than 15 variables
  let foo0 = ""
  let foo1 = ""
  let foo2 = ""
  let foo3 = ""
  let foo4 = ""
  let foo5 = ""
  let foo6 = ""
  let foo7 = ""
  let foo8 = ""
  let foo9 = ""
  let foo10 = ""
  let foo11 = ""
  let foo12 = ""
  let foo13 = ""
  let foo14 = ""
endfunction

" nvim crash <updatetime> ms after the second call Foo()
set updatetime=100
call Foo()
call Foo()
```

Please let me know if you need anything else.

**Comment by zeertzjq** [comment]
I can reproduce this. ASAN log:
```
=================================================================
==215959==ERROR: AddressSanitizer: attempting double-free on 0x619000003c80 in thread T0:
    #0 0x55ef1613eeb9 in free (**/neovim/build/bin/nvim+0xbfeeb9)
    #1 0x55ef16d12e84 in xfree **/neovim/src/nvim/memory.c:122:3
    #2 0x55ef16ac717f in hash_clear **/neovim/src/nvim/hashtab.c:60:5
    #3 0x55ef165c7252 in vars_clear_ext **/neovim/src/nvim/eval.c:9334:3
    #4 0x55ef16551a29 in vars_clear **/neovim/src/nvim/eval.c:9304:3
    #5 0x55ef167bfd77 in free_funccal_contents **/neovim/src/nvim/eval/userfunc.c:616:3
    #6 0x55ef167bf7fd in free_unref_funccal **/neovim/src/nvim/eval/userfunc.c:3341:7
    #7 0x55ef16554fa7 in garbage_collect **/neovim/src/nvim/eval.c:5389:16
    #8 0x55ef16a7d5e9 in before_blocking **/neovim/src/nvim/getchar.c:1396:5
    #9 0x55ef16fb3804 in os_inchar **/neovim/src/nvim/os/input.c:125:9
    #10 0x55ef174511c4 in state_enter **/neovim/src/nvim/state.c:53:13
    #11 0x55ef16de930a in normal_enter **/neovim/src/nvim/normal.c:469:3
    #12 0x55ef16ba9a33 in main **/neovim/src/nvim/main.c:547:3
    #13 0x7fcdf3881b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    #14 0x55ef16097f6d in _start (**/neovim/build/bin/nvim+0xb57f6d)

0x619000003c80 is located 0 bytes inside of 1024-byte region [0x619000003c80,0x619000004080)
freed by thread T0 here:
    #0 0x55ef1613eeb9 in free (**/neovim/build/bin/nvim+0xbfeeb9)
    #1 0x55ef16d12e84 in xfree **/neovim/src/nvim/memory.c:122:3
    #2 0x55ef16ac717f in hash_clear **/neovim/src/nvim/hashtab.c:60:5
    #3 0x55ef165c7252 in vars_clear_ext **/neovim/src/nvim/eval.c:9334:3
    #4 0x55ef16551a29 in vars_clear **/neovim/src/nvim/eval.c:9304:3
    #5 0x55ef16798f13 in cleanup_function_call **/neovim/src/nvim/eval/userfunc.c:640:5
    #6 0x55ef167968e3 in call_user_func **/neovim/src/nvim/eval/userfunc.c:1208:3
    #7 0x55ef1678d1fa in call_func **/neovim/src/nvim/eval/userfunc.c:1580:11
    #8 0x55ef167896e6 in get_func_tv **/neovim/src/nvim/eval/userfunc.c:459:11
    #9 0x55ef167bc391 in ex_call **/neovim/src/nvim/eval/userfunc.c:3008:9
    #10 0x55ef1689ddf9 in do_one_cmd **/neovim/src/nvim/ex_docmd.c:1983:5
    #11 0x55ef16889059 in do_cmdline **/neovim/src/nvim/ex_docmd.c:604:20
    #12 0x55ef16874454 in do_source **/neovim/src/nvim/ex_cmds2.c:2242:5
    #13 0x55ef1686f7ad in cmd_source **/neovim/src/nvim/ex_cmds2.c:1805:14
    #14 0x55ef1686fe4d in ex_source **/neovim/src/nvim/ex_cmds2.c:1786:3
    #15 0x55ef1689ddf9 in do_one_cmd **/neovim/src/nvim/ex_docmd.c:1983:5
    #16 0x55ef16889059 in do_cmdline **/neovim/src/nvim/ex_docmd.c:604:20
    #17 0x55ef16e3b640 in nv_colon **/neovim/src/nvim/normal.c:4821:18
    #18 0x55ef16e27bc5 in normal_execute **/neovim/src/nvim/normal.c:1164:3
    #19 0x55ef17451399 in state_enter **/neovim/src/nvim/state.c:68:26
    #20 0x55ef16de930a in normal_enter **/neovim/src/nvim/normal.c:469:3
    #21 0x55ef16ba9a33 in main **/neovim/src/nvim/main.c:547:3
    #22 0x7fcdf3881b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)

previously allocated by thread T0 here:
    #0 0x55ef1613f1c9 in malloc (**/neovim/build/bin/nvim+0xbff1c9)
    #1 0x55ef16d12b92 in try_malloc **/neovim/src/nvim/memory.c:74:15
    #2 0x55ef16d12d5c in xmalloc **/neovim/src/nvim/memory.c:108:15
    #3 0x55ef16ac9e13 in hash_may_resize **/neovim/src/nvim/hashtab.c:347:7
    #4 0x55ef16ac8ec1 in hash_add_item **/neovim/src/nvim/hashtab.c:233:3
    #5 0x55ef16ac8be5 in hash_add **/neovim/src/nvim/hashtab.c:212:3
    #6 0x55ef1675feb4 in tv_dict_add **/neovim/src/nvim/eval/typval.c:1713:10
    #7 0x55ef165c907a in set_var_const **/neovim/src/nvim/eval.c:9525:9
    #8 0x55ef16561c06 in set_var_lval **/neovim/src/nvim/eval.c:2433:7
    #9 0x55ef165ee0af in ex_let_one **/neovim/src/nvim/eval.c:1951:9
    #10 0x55ef16577dbe in ex_let_vars **/neovim/src/nvim/eval.c:1470:9
    #11 0x55ef16571085 in ex_let_const **/neovim/src/nvim/eval.c:1441:13
    #12 0x55ef16571176 in ex_let **/neovim/src/nvim/eval.c:1361:3
    #13 0x55ef1689ddf9 in do_one_cmd **/neovim/src/nvim/ex_docmd.c:1983:5
    #14 0x55ef16889059 in do_cmdline **/neovim/src/nvim/ex_docmd.c:604:20
    #15 0x55ef1679433c in call_user_func **/neovim/src/nvim/eval/userfunc.c:1114:5
    #16 0x55ef1678d1fa in call_func **/neovim/src/nvim/eval/userfunc.c:1580:11
    #17 0x55ef167896e6 in get_func_tv **/neovim/src/nvim/eval/userfunc.c:459:11
    #18 0x55ef167bc391 in ex_call **/neovim/src/nvim/eval/userfunc.c:3008:9
    #19 0x55ef1689ddf9 in do_one_cmd **/neovim/src/nvim/ex_docmd.c:1983:5
    #20 0x55ef16889059 in do_cmdline **/neovim/src/nvim/ex_docmd.c:604:20
    #21 0x55ef16874454 in do_source **/neovim/src/nvim/ex_cmds2.c:2242:5
    #22 0x55ef1686f7ad in cmd_source **/neovim/src/nvim/ex_cmds2.c:1805:14
    #23 0x55ef1686fe4d in ex_source **/neovim/src/nvim/ex_cmds2.c:1786:3
    #24 0x55ef1689ddf9 in do_one_cmd **/neovim/src/nvim/ex_docmd.c:1983:5
    #25 0x55ef16889059 in do_cmdline **/neovim/src/nvim/ex_docmd.c:604:20
    #26 0x55ef16e3b640 in nv_colon **/neovim/src/nvim/normal.c:4821:18
    #27 0x55ef16e27bc5 in normal_execute **/neovim/src/nvim/normal.c:1164:3
    #28 0x55ef17451399 in state_enter **/neovim/src/nvim/state.c:68:26
    #29 0x55ef16de930a in normal_enter **/neovim/src/nvim/normal.c:469:3

SUMMARY: AddressSanitizer: double-free (**/neovim/build/bin/nvim+0xbfeeb9) in free
==215959==ABORTING
```

**Comment by erw7** [comment]
This seems to be an issue that has already been fixed in vim in 8.2.1845 (https://github.com/vim/vim/commit/8d739de43b84ef7817b3b5b826d1cbfe7572a62a). 8.2.1845 is a fix for Vim9, but at least the following changes are needed for neovim as well.

```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index b27a11c24..c68b944b6 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -9332,7 +9332,7 @@ void vars_clear_ext(hashtab_T *ht, int free_val)
     }
   }
   hash_clear(ht);
-  ht->ht_used = 0;
+  hash_init(ht);
 }

 /*
```
